### PR TITLE
Feat/tray

### LINKED ISSUE #37: Adding system tray.

### PR DESCRIPTION
**Pull request by TornaxO7**
Alright, so it's possible now to create a tray.

The question is: How do we want to run this tray?
- In a forked process or
- should we let the user do that?

@eneoli 

image

Closes #37

### CODE REVIEW

**Review by eneoli — COMMENTED** [review]
lgtm. I think the only option is to spawn a thread for the tray. What do you mean by letting the user do it?

**Comment by TornaxO7** [comment]
> lgtm. I think the only option is to spawn a thread for the tray.

Hm... I'd prefer a child process than a thread.

> What do you mean by letting the user do it?

That the `flakeshot` process becomes the tray-process itself. Currently I'm spawning a thread where the tray runs but I think it's better if the user can start the tray with `flakeshot tray` and then it's up to the user how he/she wants to run it in the background. Either by just simply moving it into a background process with `flakeshot tray &` or even bring it to the init system with `flakeshot tray & disown` or writing a systemd-service file etc.

**Comment by TornaxO7** [comment]
I'm actually prefering to not create a child process and let the process, if it should invoke the system tray,
become the tray itself (for the time being). This should reduce the complexity although it may be not that user friendly.

**Comment by eneoli** [comment]
Letting the main process be the tray sounds reasonable as users will likely create new screenshots by interacting with the tray first.

The best option would be to have multiple cli commands. One could be - as you mentioned - `flakeshot tray`. But I rather prefer a different approach:

Let the default behavior (i.e. just running `flakeshot`) create and run the tray. We could use a second command like `flakeshot snap` to directly create a screenshot.


Sounds that good to you? :)

**Comment by TornaxO7** [comment]
> The best option would be to have multiple cli commands. One could be - as you mentioned - `flakeshot tray`. But I rather prefer a different approach:
> 
> Let the default behavior (i.e. just running `flakeshot`) create and run the tray. We could use a second command like `flakeshot snap` to directly create a screenshot.
> 
> Sounds that good to you? :)

Personally, it's more intuitive that the screenshot-functionality appears if I just execute `flakeshot` but `flameshot` is also starting a system tray per default if you just type `flameshot`, so we can stick to it.

**Comment by eneoli** [comment]
There also a different point to consider: the Wayland (and perhaps also X11?) clipboard contents get lost if the application that created them exits. Thus we have to let flakeshot active after copying the image into the clipboard or let the user  use some sort of clipboard daemon like `clipmon`

**Comment by TornaxO7** [comment]
That's right!

File: flake.nix [hunk 1]
```diff
@@ -84,7 +84,10 @@
                 stdenv = pkgs.stdenvAdapters.useMoldLinker pkgs.clangStdenv;
               }
               {
-                packages = [ rust-toolchain ] ++ flakeshot.nativeBuildInputs ++ flakeshot.buildInputs;
+                packages = with pkgs; [
+                  snixembed
```
Comment (eneoli) [inline flake.nix:88]: Won't this clash if some other tray is installed?

File: src/tray.rs [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    fn icon_name(&self) -> String {
+        "flakeshot-tray".into()
+    }
+
+    fn id(&self) -> String {
+        env!("CARGO_PKG_NAME").into()
+    }
+
+    fn activate(&mut self, _x: i32, _y: i32) {
+        println!("Leftclick action is still under development.");
```
Comment (eneoli) [inline src/tray.rs:16]: Why not starting the gui on click?

File: src/tray.rs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    }
+
+    fn menu(&self) -> Vec<ksni::MenuItem<Self>> {
+        use ksni::menu::*;
+        vec![SubMenu {
+            label: "System tray is under development".into(),
```
Comment (eneoli) [inline src/tray.rs:22]: Quit button?

**Review by eneoli — CHANGES_REQUESTED** [review]
see annotations

File: flake.nix (hunk 1 above)
Comment (TornaxO7) [reply flake.nix:88]: `snixembed` has to be started in the cli. So as long you don't start it, nothing happens.

File: src/tray.rs (hunk 2 above)
Comment (TornaxO7) [reply src/tray.rs:16]: Fair point.

File: src/tray.rs (hunk 3 above)
Comment (TornaxO7) [reply src/tray.rs]: added

File: src/lib.rs [hunk 4]
```diff
@@ -1,7 +1,12 @@
 //! Welcome to the code-documentation of flakeshot!
+
+use frontend::main_window::AppModel;
+use gtk4::CssProvider;
+use relm4::RelmApp;
```
Comment (eneoli) [inline src/lib.rs:5]:
```suggestion
use relm4::RelmApp;

```

**Review by eneoli — CHANGES_REQUESTED** [review]

File: src/tray.rs [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+            label: "Quit".into(),
+            activate: Box::new(|_| std::process::exit(0)),
+            ..Default::default()
+        }
+        .into()]
+    }
+}
+
+pub fn start() {
+    let _ = ksni::spawn(Tray).unwrap();
```
Comment (eneoli) [inline src/tray.rs:31]: remove the anonymous var

Comment (TornaxO7) [reply src/tray.rs:31]: done

**Review by eneoli — APPROVED** [review]

**Pull request merged**